tests/test_bms_transformations.py

```diff
@@ -0,0 +1,450 @@
+    v2 = np.array([-4, 5, -6]) * 1e-2
+
+    L1 = bms_transformations.Lorentz_transformation(rotation=q1.components, boost=v1)
+    L2 = bms_transformations.Lorentz_transformation(rotation=q2.components, boost=v2)
+
+    q1v1q2v2 = np.linalg.multi_dot([boost_matrix(v2), rotation_matrix(q2), boost_matrix(v1), rotation_matrix(q1)])
+    q2v2q1v1 = np.linalg.multi_dot([boost_matrix(v1), rotation_matrix(q1), boost_matrix(v2), rotation_matrix(q2)])
+
+    L1L2 = L1.compose(L2)
+    L1L2_matrix = np.linalg.multi_dot([boost_matrix(L1L2.boost), rotation_matrix(L1L2.rotation)])
+
+    L2L1 = L2.compose(L1)
+    L2L1_matrix = np.linalg.multi_dot([boost_matrix(L2L1.boost), rotation_matrix(L2L1.rotation)])
+
+    assert np.allclose(q1v1q2v2, L1L2_matrix)
+    assert np.allclose(q2v2q1v1, L2L1_matrix)
+
+
+def test_Lorentz_abd_inverse():
+    mass = 2.0
+    spin = 0.456
+    ell_max = 8
+    u = np.linspace(-100, 100, num=500)
+    psi2, psi1, psi0 = kerr_schild(mass, spin, ell_max)
+    abd = ABD.from_initial_values(u, ell_max=ell_max, psi2=psi2, psi1=psi1)
+
+    q = np.quaternion(1, 2, 3, 4).normalized()
+    v = np.array([1, 2, 3]) * 1e-4
+
+    Lorentz = bms_transformations.Lorentz_transformation(rotation=q, boost=v, order=["rotation", "boost"])
+
+    Lorentz_inv = Lorentz.inverse(output_order=["rotation", "boost"])
+
+    abd_prime = abd.transform(frame_rotation=Lorentz.rotation.components, boost_velocity=Lorentz.boost)
+
+    abd_check = abd_prime.transform(frame_rotation=Lorentz_inv.rotation.components, boost_velocity=Lorentz_inv.boost)
+
+    abd_interp = abd.interpolate(abd_check.t)
+
+    assert np.allclose(
+        np.array(
+            [abd_interp.sigma, abd_interp.psi4, abd_interp.psi3, abd_interp.psi2, abd_interp.psi1, abd_interp.psi0]
+        ),
+        np.array([abd_check.sigma, abd_check.psi4, abd_check.psi3, abd_check.psi2, abd_check.psi1, abd_check.psi0]),
+    )
+
+
+def test_Lorentz_abd_composition():
+    mass = 2.0
+    spin = 0.456
+    ell_max = 8
+    u = np.linspace(-100, 100, num=500)
+    psi2, psi1, psi0 = kerr_schild(mass, spin, ell_max)
+    abd = ABD.from_initial_values(u, ell_max=ell_max, psi2=psi2, psi1=psi1)
+
+    q1 = np.quaternion(1, 2, 3, 4).normalized()
+    v1 = np.array([1, 2, 3]) * 1e-4
+    q2 = np.quaternion(5, -6, 7, -8).normalized()
+    v2 = np.array([-4, 5, -6]) * 1e-4
+
+    Lorentz1 = bms_transformations.Lorentz_transformation(rotation=q1, boost=v1, order=["rotation", "boost"])
+
+    Lorentz2 = bms_transformations.Lorentz_transformation(rotation=q2, boost=v2, order=["rotation", "boost"])
+
+    Lorentz_composed = Lorentz2.compose(Lorentz1)
+
+    abd1 = abd.transform(frame_rotation=Lorentz1.rotation.components, boost_velocity=Lorentz1.boost)
+
+    abd2 = abd1.transform(frame_rotation=Lorentz2.rotation.components, boost_velocity=Lorentz2.boost)
+
+    abd_composed = abd.transform(
+        frame_rotation=Lorentz_composed.rotation.components, boost_velocity=Lorentz_composed.boost
+    )
```

Comment:
See, this is tricky.  It says that applying `L1` to `abd`, then applying `L2` to the transformed `abd` is the same as applying `L1 * L2`.  Put another way: if we think of `L1` and `L2` as *operators* on the waveform, we would write that as `L2(L1(abd)) = (L1*L2)(abd)`.  So composition is opposite to multiplication: `L1*L2 = L2∘L1`

The reason, of course, is that `L2` is expressed in the first frame, so to apply it to `L1(abd)`, you have to transform it into the second frame — which is `L1*L2*L1.inverse()` — so `L2∘L1 = L1*L2*L1.inverse()*L1 = L1*L2`.

This is the subtlety I was worried about in our phone call yesterday.  And I agree that you've handled it correctly here, but just reading the documentation of `compose`, I think it would be easy to order things the wrong way around.

Reply:
I'll try and clean up the documentation. I also wrote up some TeX that Leo edited that we think would be useful to add to the documentation of `scri` to make things clearer (https://www.overleaf.com/6964495134vvzdnkwnwkhv)

Reply:
We should continue adding details here — namely, we didn't include the math of conjugating group elements by other elements for the composition of two BMS elements.

Reply:
I just updated the Overleaf with that abstract calculation, but not the details of how conjugating works.